src/design_notes/general_coroutines.md

```diff
@@ -0,0 +1,258 @@
+# Generalized coroutines
+
+Since even before Rust 1.0, users have desired the ability to `yield` like in
+other languages. The compiler infrastructure to achieve this, along with an
+unstable syntax, have existed for a while now. But despite *a lot* of debate,
+we've failed to polish the feature up enough to stabilize it. I've tried to
+write up a summary of the different design considerations and the past debate
+around them below:
+
+## Coroutines vs Generators
+
+- The distinction between a "coroutine" and a "generator" can be a bit vague,
+  varying from one discussion to the next.
+- In these notes a generator is anything which *directly* implements `Iterator`
+  or `Stream` while a coroutine is anything which can take arbitrary input,
+  yield arbitrary output, and later resume execution at the previous yield.
+- Thus, the "generator" syntax proposed in [eRFC-2033][1] and currently
+  implemented behind the "generator" feature is actually a coroutine syntax for
+  the sake of these notes, *not a true generator*.
+- Note also that "coroutines" here are really "semicoroutines" since they can
+  only yield back to their caller.
+- Generators are a specialization of coroutines.
+- I will continue to group the [original eRFC text][1] and the later [generator
+  resume arguments](https://github.com/rust-lang/rust/pull/68524) extension
+  togther as "[eRFC-2033][1]". That way I only have 3 big proposals to deal
+  with.
+
+```rust
+// This is a coroutine
+|stuff| {
+  yield stuff + 1;
+  yield stuff + 2;
+  return stuff + 3;
+}
+
+// This is a generator
+stream! {
+  yield 1;
+  yield 2;
+  yield 3;
+}
+```
+
+## Coroutine trait
+
+- The coroutine syntax must produce implementations of some trait.
+- [RFC-2781][2] and [eRFC-2033][1] propose the `Generator` trait.
+- Note that Rust's coroutines and subroutines look the same from the outside:
+  take input, mutate state, produce output.
+- Thus, [MCP-49][3] proposes using the `Fn*` traits instead, including a new
+  `FnPin` for immovable coroutines.
+  - Hierarchy: `Fn` is `FnMut + Unpin` is `FnPin` is `FnOnce`.
+    - May not be *required* at the trait level (someone may someday find a use
+      to implementing `FnMut + !FnPin`) but all closures implement the traits in
+      this order.
+
+## Coroutine syntax
+
+- The closure syntax is reused for coroutines by [eRFC-2033][1], [RFC-2781][2],
+  and [MCP-49][3].
+- Commentators have suggested that the differences between coroutines and
+  closures under [eRFC-2033][1] and [RFC-2781][2] justify an entirely distinct
+  syntax to reduce confusion.
+- [MCP-49][3] fully reuses the *semantics* of closures, greatly simplifying the
+  design space and making the shared syntax obvious.
+
+## Taking input
```

Comment:
There's another issue WRT input that I don't see discussed here, which is how to shape the trait such that input can have a limited or unlimited lifetime. For example, a generator might want to take in input that is a reference to a buffer which is rewritten to on every `yield`, yet the type of each input would therefore be different because it depends on the lifetime of the input argument. Other generators might want to ask for input which is independent of lifetime so that the input can continue to be referenced across multiple `yield`s. To give an example, it's not clear whether or not the following should compile:

```
let mut my_generator = gen |x: &i32| {
  let mut v: Vec<&i32> = Vec::new();
  v.push(x);
  for i in 0..128 {
    v.push(x);
  }
};
```

If we add `let mut a = 5; my_generator.resume(&a); a = 6; my_generator.resume(&a);`, it becomes clear that *something* shouldn't compile.

Reply:
This is a bit of a tricky one and I'm not sure how deep do go in these notes. RFC-2781 leaves it in unresolved questions with a note that they might restrict inputs to `'static` types. Since MCP-49 acts as much like closures as possible, and treats the witness and capture data the same whenever possible, pushing a reference into cross-yield (witnessed) state would act like
```
let mut state = Vec::new();
move |x: &i32| state.push(x)
               ^^^^^^^^^^^^^ `x` escapes the closure body here
```
does today and would result in a similar "borrowed data escapes into closure state" error. Ideally someday we'd do something nicer but any such solution would apply to both captured state and witnessed state in the same way.

Your exact example would actually be allowed since `v` is dropped before any yield occurs but I'm assuming that you intended to put a `yield` into the for loop. I also want to point out that this problem wouldn't apply to any planned `gen fn`/`gen { }` syntax since, as I mentioned in Niko's review above, smart generator proposals generally avoid taking resume arguments. Being super technical, it only applies to *coroutine* proposals.

Does that clear this up? I'll copy that explanation int the  bullet points if you think it makes sufficient sense.

Reply:
Yes, I did intend to include a `yield` in the loop, thanks for the clarification!